x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/DataNodeRequestSender.java

```diff
@@ -0,0 +1,336 @@
+            }
+        }, listener::onFailure));
+    }
+
+    private void trySendingRequestsForPendingShards(TargetShards targetShards, ComputeListener computeListener) {
+        changed.set(true);
+        final ActionListener<Void> listener = computeListener.acquireAvoid();
+        try {
+            while (sendingLock.tryLock()) {
+                try {
+                    if (changed.compareAndSet(true, false) == false) {
+                        break;
+                    }
+                    for (ShardId shardId : pendingShardIds) {
+                        if (targetShards.getShard(shardId).remainingNodes.isEmpty()) {
+                            final Exception failure = shardFailures.get(shardId);
+                            if (failure == null) {
+                                shardFailures.put(shardId, new ShardNotFoundException(shardId, "no shard copies found {}", shardId));
+                            }
+                            aborted.set(true);
+                        }
+                    }
+                    if (aborted.get()) {
+                        for (Exception e : shardFailures.values()) {
+                            computeListener.acquireAvoid().onFailure(e);
+                        }
+                    } else {
+                        var nodeRequests = selectNodeRequests(targetShards);
+                        for (NodeRequest request : nodeRequests) {
+                            sendOneNodeRequest(targetShards, computeListener, request);
+                        }
+                    }
+                } finally {
+                    sendingLock.unlock();
+                }
+            }
+        } finally {
+            listener.onResponse(null);
+        }
+    }
+
+    private void sendOneNodeRequest(TargetShards targetShards, ComputeListener computeListener, NodeRequest request) {
+        final ActionListener<List<DriverProfile>> listener = computeListener.acquireCompute();
+        sendRequest(request.node, request.shardIds, request.aliasFilters, new NodeListener() {
+            void onAfter(List<DriverProfile> profiles) {
+                nodePermits.get(request.node).release();
+                trySendingRequestsForPendingShards(targetShards, computeListener);
+                listener.onResponse(profiles);
+            }
+
+            @Override
+            public void onResponse(DataNodeComputeResponse response) {
+                // remove failures of successful shards
+                for (ShardId shardId : targetShards.shardIds()) {
+                    if (shardFailures.containsKey(shardId) == false) {
+                        shardFailures.remove(shardId);
+                    }
+                }
+                for (Map.Entry<ShardId, Exception> e : response.shardLevelFailures().entrySet()) {
+                    final ShardId shardId = e.getKey();
+                    shardFailures.compute(shardId, (k, v) -> mergeShardExceptions(v, e.getValue()));
```

Comment:
This is safe concurrently because:
1. You are using compute
2. More fundamentally, you don't have more than one request out for each individual ShardId at a time.

Reply:
Worth a comment for folks like me.

Reply:
There is an issue as it might concurrently suppress the same exception. I pushed https://github.com/elastic/elasticsearch/pull/120774/commits/c010fdc291f791afca8009ac6cd3ce2446d50573